lib/initialConfigValidation.js

```diff
@@ -0,0 +1,89 @@
+'use strict'
+
+const Ajv = require('ajv')
+const ajv = new Ajv({ removeAdditional: true, useDefaults: true, coerceTypes: true })
+
+// We add a keyword that allow us to set default values
+ajv.addKeyword('setDefaultValue', {
+  modifying: true,
+  validate: function (schemaParamValue, validatedParamValue, validationSchemaObject, currentDataPath, validatedParamObject, validatedParam) {
+    validatedParamObject[validatedParam] = schemaParamValue
+
+    return true
+  },
+  errors: false
+})
+
+const optsSchema = {
+  type: 'object',
+  additionalProperties: false,
+  properties: {
+    bodyLimit: { type: 'integer' },
+    caseSensitive: { type: 'boolean' },
+    http2: {
+      oneOf: [
+        { type: 'boolean' },
+        { type: 'null' }
+      ]
+    },
+    https: {
+      if: {
+        not: {
+          oneOf: [
+            { type: 'boolean' },
+            { type: 'null' },
+            {
+              type: 'object',
+              additionalProperties: false,
+              required: ['allowHTTP1'],
+              properties: {
+                allowHTTP1: { type: 'boolean' }
+              }
+            }
+          ]
+        }
+      },
+      then: { setDefaultValue: true }
+    },
+    ignoreTrailingSlash: { type: 'boolean' },
+    maxParamLength: { type: 'integer' },
+    onProtoPoisoning: { type: 'string' },
+    pluginTimeout: { type: 'integer' },
+    requestIdHeader: { type: 'string' }
+  }
+}
+
+const validate = ajv.compile(optsSchema)
+
+function validateInitialConfig (options) {
+  // We clone the initial configuration options object
+  let opts = Object.assign({}, options)
+
+  const isOptionValid = validate(opts)
+
+  if (!isOptionValid) {
+    const error = new Error(validate.errors.map(e => e.message))
+    error.errors = validate.errors
+    throw error
+  }
+
+  return deepFreezeObject(opts)
+
+  function deepFreezeObject (object) {
+    let properties = Object.getOwnPropertyNames(object)
+
+    for (const name of properties) {
+      let value = object[name]
+
+      if (typeof value !== 'object' || Buffer.isBuffer(value)) {
+        continue
+      }
+
+      object[name] = value && typeof value === 'object' ? deepFreezeObject(value) : value
```

Comment:
Deep freezing the options object without deep cloning it could cause values in the original object to be frozen:

```js
'use strict'

const Fastify = require('fastify')
const fs = require('fs')
const originalOptions = {
  https: {
    allowHTTP1: true,
    key: fs.readFileSync(path.join(__dirname, '..', 'https', 'fastify.key')),
    cert: fs.readFileSync(path.join(__dirname, '..', 'https', 'fastify.cert'))
  }
}

Object.isFrozen(originalOptions) // false
Object.isFrozen(originalOptions.https) // false

const fastify = Fastify(originalOptions)

Object.isFrozen(originalOptions) // false
Object.isFrozen(originalOptions.https) // true <-- object was modified
```

This is an edge case and probably won't affect anyone, but cloning before recursing here would be a simple way to solve this:

```js
deepFreezeObject(Object.assign({}, value))
```

Reply:
Hi.
Nice catch :smiley: 
I made the change you suggested and added one more test to address this behavior ^^
Thank you :+1: